extensions.go

```diff
@@ -2,13 +2,83 @@ package csproto
 
 import (
 	"fmt"
+	"reflect"
 
 	gogo "github.com/gogo/protobuf/proto"
 	google "github.com/golang/protobuf/proto" //nolint: staticcheck // we're using this deprecated package intentionally
 	googlev2 "google.golang.org/protobuf/proto"
 	"google.golang.org/protobuf/reflect/protoreflect"
 )
 
+// Extension is a combined struct describing features from gogo, google v1 & v2, containing all
+// common descriptor values and the underlying extension value.
+// NOTE: Certain tags have been excluded as they are deprecated or non-existent or certain libraries
+type Extension struct {
+	ExtensionType interface{}
+	Field         int32
+	Name          string
+	Filename      string
+	Value         interface{}
+}
+
+// RangeExtensions iterates through all extension descriptors of a given proto message, calling fn
+// on each iteration. It returns immediately on any error encountered.
+func RangeExtensions(msg interface{}, fn func(ext Extension) error) error {
+	value := reflect.ValueOf(msg)
+	msgType := deduceMsgType(msg, value.Type())
```

Comment:
Use `MsgType()` instead of `deduceMsgType()` to take advantage of the cache.  `deduceMsgType` does the full type analysis on `msg` in every call.
https://github.com/CrowdStrike/csproto/blob/main/message_types.go#L28